start adding part one content (#12035)
flutter/website: src/content/tutorial/ui/6-stateful-widget.md

---
title: Stateful widgets
description: Learn about StatefulWidgets and rebuilding Flutter UI.
permalink: /tutorial/stateful-widget/
---

{%- comment %} TODO(ewindmill) embed video {%- endcomment %}

So far, your app displays a grid and an input field, but the grid
doesn't yet update to reflect the user’s guesses. When this app is
complete, each tile in the next unfilled row should update after each
submitted user guess by:

* Displaying the correct letter  
* Changing color to reflect whether the letter is correct (green), is
  in the word but at an incorrect position (yellow), or incorrect in
  both position and type (grey).

To handle this dynamic behavior, you need to convert `GamePage` from a
`StatelessWidget` to a [`StatefulWidget`][].

## Why stateful widgets?

When a widget's appearance or data needs to change during its
 lifetime, you need a `StatefulWidget` and a companion `State` object.
While the `StatefulWidget` itself is still immutable (its properties
can't change after creation), the `State` object is long-lived, can
hold mutable data, and can be rebuilt when that data changes, causing
the UI to update.

{%- comment %} TODO(ewindmill) diagram {%- endcomment %}

> The basic `StatefulWidget` structure:

Reviewer: ```suggestion
Here is the basic `StatefulWidget` structure (don't do anything yet):
```

```dart
class ExampleWidget extends StatefulWidget {
  ExampleWidget({super.key});
  
  @override
  State<ExampleWidget> createState() => _ExampleWidgetState();
}

class _ExampleWidgetState extends State<ExampleWidget> {
  @override
  Widget build(BuildContext context) {
   return Container();
  }
}
```

## Convert `GamePage` to a stateful widget

> To convert the `GamePage` widget (or any other) from a stateless widget to a stateful widget, do the following steps:

Reviewer: ```suggestion
To convert the `GamePage` widget (or any other) from
a stateless widget to a stateful widget, do the following steps:
```

1. Change `GamePage` to extend `StatefulWidget` instead of

>    `StatelessWidget`.  

Reviewer: Wait! Wait! Wait! I distinctly remember that you can do this automatically in an IDE. Is that no longer true?

---
title: Stateful widgets
description: Learn about StatefulWidgets and rebuilding Flutter UI.
permalink: /tutorial/stateful-widget/
---

{%- comment %} TODO(ewindmill) embed video {%- endcomment %}

So far, your app displays a grid and an input field, but the grid
doesn't yet update to reflect the user’s guesses. When this app is
complete, each tile in the next unfilled row should update after each
submitted user guess by:

* Displaying the correct letter.
* Changing color to reflect whether the letter is correct (green), is

>   in the word but at an incorrect position (yellow), or incorrect in
>   both position and type (grey).

Reviewer: ```suggestion
  in the word but at an incorrect position (yellow), or doesn't appear
  in the word at all (grey).
```

To handle this dynamic behavior, you need to convert `GamePage` from a
`StatelessWidget` to a [`StatefulWidget`][].

## Why stateful widgets?

When a widget's appearance or data needs to change during its
 lifetime, you need a `StatefulWidget` and a companion `State` object.
While the `StatefulWidget` itself is still immutable (its properties
can't change after creation), the `State` object is long-lived, can
hold mutable data, and can be rebuilt when that data changes, causing
the UI to update.

> This widget tree imagines a simple app that has a counter, 
> which increases when the button is pressed.

Reviewer: ```suggestion
For example, the following widget tree imagines a simple app
that has a counter that increases when the button is pressed,
and uses a stateful widget.
```

<img src='/assets/images/docs/tutorial/widget_tree_stateful.png' alt="A diagram of a widget tree with a stateful widget and state object.">

Here is the basic `StatefulWidget` structure (don't do anything yet):

```dart
class ExampleWidget extends StatefulWidget {
  ExampleWidget({super.key});
  
  @override
  State<ExampleWidget> createState() => _ExampleWidgetState();
}

class _ExampleWidgetState extends State<ExampleWidget> {
  @override
  Widget build(BuildContext context) {
   return Container();
  }
}
```

## Convert `GamePage` to a stateful widget

To convert the `GamePage` widget (or any other) from
a stateless widget to a stateful widget, do the following steps:

1. Change `GamePage` to extend `StatefulWidget` instead of
   `StatelessWidget`.  
2. Create a new class named `_GamePageState`, that extends
   `State<GamePage>`. This new class will hold the mutable state and
   the `build` method. Move the `build` method and all properties
   *instantiated on the widget*  from `GamePage` to the state object.  
3. Implement the `createState()` method in `GamePage`, which returns
   an instance of `_GamePageState`. 

:::tip Quick assists

> You don't have to manually do this work, as the Flutter plugin for
> VSCode and Intellij provides ["quick assists"][], which will do this

Reviewer: ```suggestion
You don't have to manually do this work, as the Flutter plugins for
VS Code and IntelliJ provides ["quick assists"][], which will do this
```
